### PR TITLE
Cleaned up some code from gameserver class on bootstrap method.

### PR DESCRIPTION
**Pull request by TheMeinerLP**
Split some massiv code into smaller methodes from massiv bootstrap method.
And also add java documentation to the splitted methods. 
A rework is planned to the future of 2.2.0/1. This is only a pr for improve legibility of the class.

### CODE REVIEW

File: cloudnet-wrapper/src/main/java/de/dytanic/cloudnetwrapper/server/GameServer.java [hunk 1]
```diff
@@ -387,17 +315,126 @@ public boolean bootstrap() throws Exception {
...
+    /**
+     * Configure a server like not a glowstone server type.
+     *
+     * @return Return the finish Server info.
+     */
+    private ServerInfo configureNoGlowstoneServer() {
+        String motd;
+        int maxPlayers;
+        Properties properties = new Properties();
+        try (InputStreamReader inputStreamReader = new InputStreamReader(Files.newInputStream(Paths.get(path + "/server.properties")))) {
```
Comment (GiantTreeLP) [inline cloudnet-wrapper/src/main/java/de/dytanic/cloudnetwrapper/server/GameServer.java]: Use the separator-less version of `Paths.get` by specifying each directory separately
```suggestion
        try (InputStreamReader inputStreamReader = new InputStreamReader(Files.newInputStream(Paths.get(path,  "server.properties")))) {
```

File: cloudnet-wrapper/src/main/java/de/dytanic/cloudnetwrapper/server/GameServer.java [hunk 2]
```diff
@@ -387,17 +315,126 @@ public boolean bootstrap() throws Exception {
...
-        serverProcess.setServerStage(ServerStage.PROCESS);
-        CloudNetWrapper.getInstance().getServers().put(this.serverProcess.getMeta().getServiceId().getServerId(), this);
-        serverProcess.setServerStage(ServerStage.NET_INIT);
-        return true;
+    /**
+     * Configure a server like not a glowstone server type.
+     *
+     * @return Return the finish Server info.
+     */
+    private ServerInfo configureNoGlowstoneServer() {
```
Comment (GiantTreeLP) [inline cloudnet-wrapper/src/main/java/de/dytanic/cloudnetwrapper/server/GameServer.java]: Choose a more fitting name, this code does not deal with Glowstone, so Glowstone has no place in the name.

File: cloudnet-wrapper/src/main/java/de/dytanic/cloudnetwrapper/server/GameServer.java [hunk 3]
```diff
@@ -387,17 +315,126 @@ public boolean bootstrap() throws Exception {
...
+    }
 
-        serverProcess.setServerStage(ServerStage.PROCESS);
-        CloudNetWrapper.getInstance().getServers().put(this.serverProcess.getMeta().getServiceId().getServerId(), this);
-        serverProcess.setServerStage(ServerStage.NET_INIT);
-        return true;
+    /**
+     * Configure a server like not a glowstone server type.
+     *
+     * @return Return the finish Server info.
```
Comment (GiantTreeLP) [inline cloudnet-wrapper/src/main/java/de/dytanic/cloudnetwrapper/server/GameServer.java]:
```suggestion
     * @return the finished server info of the configured server.
```

File: cloudnet-wrapper/src/main/java/de/dytanic/cloudnetwrapper/server/GameServer.java [hunk 4]
```diff
@@ -387,17 +315,126 @@ public boolean bootstrap() throws Exception {
...
+                              this.serverProcess.getMeta().getServerConfig(),
+                              serverProcess.getMeta().getTemplate());
+    }
 
-        serverProcess.setServerStage(ServerStage.PROCESS);
-        CloudNetWrapper.getInstance().getServers().put(this.serverProcess.getMeta().getServiceId().getServerId(), this);
-        serverProcess.setServerStage(ServerStage.NET_INIT);
-        return true;
+    /**
+     * Configure a server like not a glowstone server type.
```
Comment (GiantTreeLP) [inline cloudnet-wrapper/src/main/java/de/dytanic/cloudnetwrapper/server/GameServer.java]: This has nothing to do with Glowstone, focus on what this method does, not what it does not.

File: cloudnet-wrapper/src/main/java/de/dytanic/cloudnetwrapper/server/GameServer.java [hunk 5]
```diff
@@ -387,17 +315,126 @@ public boolean bootstrap() throws Exception {
...
+            Configuration section = configuration.getSection("server");
+            section.set("ip", CloudNetWrapper.getInstance().getWrapperConfig().getInternalIP());
+            section.set("port", serverProcess.getMeta().getPort());
+
+            maxPlayers = section.getInt("max-players");
+            motd = section.getString("motd");
+
+            configuration.set("server", section);
+            configuration.set("console.use-jline", false);
+            try (OutputStreamWriter outputStreamWriter = new OutputStreamWriter(Files.newOutputStream(Paths.get(path + "/config/glowstone.yml")),
```
Comment (GiantTreeLP) [inline cloudnet-wrapper/src/main/java/de/dytanic/cloudnetwrapper/server/GameServer.java]: You can use the `Writer` interface here

File: cloudnet-wrapper/src/main/java/de/dytanic/cloudnetwrapper/server/GameServer.java [hunk 6]
```diff
@@ -387,17 +315,126 @@ public boolean bootstrap() throws Exception {
...
+        if ((properties.isEmpty() || !properties.containsKey("max-players"))) {
+            properties.setProperty("max-players", "100");
+            FileUtility.insertData("files/server.properties", path + "/server.properties");
+            try (InputStreamReader inputStreamReader = new InputStreamReader(Files.newInputStream(Paths.get(path + "/server.properties")))) {
+                properties.load(inputStreamReader);
+            } catch (IOException e) {
+                e.printStackTrace();
+            }
+            if (this.serverGroup.getGroupMode() == ServerGroupMode.STATIC || this.serverGroup.getGroupMode() == ServerGroupMode.STATIC_LOBBY) {
+                System.err.println("Filled empty server.properties (or missing \"max-players\" entry) of server [" + this.serverProcess.getMeta()
```
Comment (GiantTreeLP) [inline cloudnet-wrapper/src/main/java/de/dytanic/cloudnetwrapper/server/GameServer.java]: Either use the logger here or ask yourself whether this is actually an error.
This situation is perfectly fine and I think the server can handle incomplete properties.

**Review by GiantTreeLP — CHANGES_REQUESTED** [review]
I don't need to write anything about your English...
Note that the changes I requested do not exclude similar issues, I just noted each change once instead of everywhere.

File: cloudnet-wrapper/src/main/java/de/dytanic/cloudnetwrapper/server/GameServer.java (hunk 5 above)
Comment (TheMeinerLP) [reply cloudnet-wrapper/src/main/java/de/dytanic/cloudnetwrapper/server/GameServer.java]: Comming soon :)

File: cloudnet-wrapper/src/main/java/de/dytanic/cloudnetwrapper/server/GameServer.java [hunk 7]
```diff
@@ -130,6 +131,100 @@ public Path getDir() {
         return dir;
     }
 
+    /**
+     * Download the {@link ServerInstallablePlugin] an copy to a web plugin cache.
+     *
+     * @param plugin The ServerInstallable plugin to download it.
+     */
+    private void downloadInstallablePlugin(ServerInstallablePlugin plugin) {
+        switch (plugin.getPluginResourceType()) {
```
Comment (GiantTreeLP) [inline cloudnet-wrapper/src/main/java/de/dytanic/cloudnetwrapper/server/GameServer.java]: You are allowed to use `if` statements here, the `switch` makes this part harder to read.

File: cloudnet-wrapper/src/main/java/de/dytanic/cloudnetwrapper/server/GameServer.java [hunk 8]
```diff
@@ -130,6 +131,100 @@ public Path getDir() {
...
+                break;
+        }
+    }
+
+    /**
+     * Copy the cloudnet api from wrapper to server folder.
+     */
+    private void copyCloudNetApi() {
+        try {
+            Files.deleteIfExists(Paths.get(path + "/plugins/CloudNetAPI.jar"));
```
Comment (GiantTreeLP) [inline cloudnet-wrapper/src/main/java/de/dytanic/cloudnetwrapper/server/GameServer.java]: Use the varargs to your advantage
```suggestion
            Files.deleteIfExists(Paths.get(path, "plugins", "CloudNetAPI.jar"));
```

File: cloudnet-wrapper/src/main/java/de/dytanic/cloudnetwrapper/server/GameServer.java [hunk 9]
```diff
@@ -130,6 +131,100 @@ public Path getDir() {
...
+            e.printStackTrace();
+        }
+    }
+
+    /**
+     * Copy some config files for different server types.
+     */
+    private void copyConfigurations() {
+        if (serverGroup.getServerType().equals(ServerGroupType.GLOWSTONE)) {
+            if (!Files.exists(Paths.get(path + "/config"))) {
```
Comment (GiantTreeLP) [inline cloudnet-wrapper/src/main/java/de/dytanic/cloudnetwrapper/server/GameServer.java]:
```suggestion
            if (!Files.exists(Paths.get(path, "config"))) {
```

File: cloudnet-wrapper/src/main/java/de/dytanic/cloudnetwrapper/server/GameServer.java [hunk 10]
```diff
@@ -130,6 +131,100 @@ public Path getDir() {
...
+                try {
+                    Files.createDirectories(Paths.get(path + "/config"));
```
Comment (GiantTreeLP) [inline cloudnet-wrapper/src/main/java/de/dytanic/cloudnetwrapper/server/GameServer.java]:
```suggestion
                    Files.createDirectories(Paths.get(path, "config"));
```

File: cloudnet-wrapper/src/main/java/de/dytanic/cloudnetwrapper/server/GameServer.java [hunk 11]
```diff
@@ -130,6 +131,100 @@ public Path getDir() {
...
+                } catch (IOException e) {
+                    e.printStackTrace();
+                }
+            }
+            if (!Files.exists(Paths.get(path + "/config/glowstone.yml"))) {
```
Comment (GiantTreeLP) [inline cloudnet-wrapper/src/main/java/de/dytanic/cloudnetwrapper/server/GameServer.java]:
```suggestion
            if (!Files.exists(Paths.get(path, "config", "glowstone.yml"))) {
```

File: cloudnet-wrapper/src/main/java/de/dytanic/cloudnetwrapper/server/GameServer.java [hunk 12]
```diff
@@ -130,6 +131,100 @@ public Path getDir() {
...
+                FileUtility.insertData("files/glowstone.yml", path + "/config/glowstone.yml");
+            }
+        }
+
+        if (!Files.exists(Paths.get(path + "/server.properties"))) {
```
Comment (GiantTreeLP) [inline cloudnet-wrapper/src/main/java/de/dytanic/cloudnetwrapper/server/GameServer.java]:
```suggestion
        if (!Files.exists(Paths.get(path, "server.properties"))) {
```

File: cloudnet-wrapper/src/main/java/de/dytanic/cloudnetwrapper/server/GameServer.java [hunk 13]
```diff
@@ -130,6 +131,100 @@ public Path getDir() {
...
+            FileUtility.insertData("files/server.properties", path + "/server.properties");
+        }
+
+        if (!Files.exists(Paths.get(path + "/bukkit.yml"))) {
```
Comment (GiantTreeLP) [inline cloudnet-wrapper/src/main/java/de/dytanic/cloudnetwrapper/server/GameServer.java]:
```suggestion
        if (!Files.exists(Paths.get(path, "bukkit.yml"))) {
```

File: cloudnet-wrapper/src/main/java/de/dytanic/cloudnetwrapper/server/GameServer.java [hunk 14]
```diff
@@ -130,6 +131,100 @@ public Path getDir() {
...
+            FileUtility.insertData("files/bukkit.yml", path + "/bukkit.yml");
+        }
+
+        if (!Files.exists(Paths.get(path + "/spigot.yml"))) {
```
Comment (GiantTreeLP) [inline cloudnet-wrapper/src/main/java/de/dytanic/cloudnetwrapper/server/GameServer.java]:
```suggestion
        if (!Files.exists(Paths.get(path, "spigot.yml"))) {
```

File: cloudnet-wrapper/src/main/java/de/dytanic/cloudnetwrapper/server/GameServer.java [hunk 15]
```diff
@@ -263,116 +272,37 @@ public boolean bootstrap() throws Exception {
...
-            }
-
-            Enumeration<Object> enumeration = this.serverProcess.getMeta().getProperties().keys();
-            while (enumeration.hasMoreElements()) {
-                String x = enumeration.nextElement().toString();
-                properties.setProperty(x, this.serverProcess.getMeta().getProperties().getProperty(x));
-            }
+            this.serverInfo = configureNormalServer();
+        } else {
+            this.serverInfo = configureNonNormalServer();
```
Comment (GiantTreeLP) [inline cloudnet-wrapper/src/main/java/de/dytanic/cloudnetwrapper/server/GameServer.java]: You are free to create proper methods for all server group types.

File: cloudnet-wrapper/src/main/java/de/dytanic/cloudnetwrapper/server/GameServer.java [hunk 16]
```diff
@@ -263,116 +272,37 @@ public boolean bootstrap() throws Exception {
...
-                                         this.serverProcess.getMeta().getServerConfig(),
-                                         serverProcess.getMeta().getTemplate());
+        serverProcess.setServerStage(ServerStage.PROCESS);
+        CloudNetWrapper.getInstance().getServers().put(this.serverProcess.getMeta().getServiceId().getServerId(), this);
+        serverProcess.setServerStage(ServerStage.NET_INIT);
+        return true;
+    }
 
+    private void generateCloudNetConfigurations() {
         if (!Files.exists(Paths.get(path + "/CLOUD"))) {
```
Comment (GiantTreeLP) [inline cloudnet-wrapper/src/main/java/de/dytanic/cloudnetwrapper/server/GameServer.java]:
```suggestion
        if (!Files.exists(Paths.get(path, "CLOUD"))) {
```

File: cloudnet-wrapper/src/main/java/de/dytanic/cloudnetwrapper/server/GameServer.java [hunk 17]
```diff
@@ -263,116 +272,37 @@ public boolean bootstrap() throws Exception {
...
-            Files.createDirectory(Paths.get(path + "/CLOUD"));
+            try {
+                Files.createDirectory(Paths.get(path + "/CLOUD"));
```
Comment (GiantTreeLP) [inline cloudnet-wrapper/src/main/java/de/dytanic/cloudnetwrapper/server/GameServer.java]:
```suggestion
                Files.createDirectory(Paths.get(path, "CLOUD"));
```

File: cloudnet-wrapper/src/main/java/de/dytanic/cloudnetwrapper/server/GameServer.java [hunk 18]
```diff
@@ -387,17 +317,131 @@ public boolean bootstrap() throws Exception {
                                                                                                                                            .get(
                                                                                                                                                path + "/CLOUD/connection.json"));
```
Comment (GiantTreeLP) [inline cloudnet-wrapper/src/main/java/de/dytanic/cloudnetwrapper/server/GameServer.java]:
```suggestion
                                                                                                                                           .get(
                                                                                                                                               path, "CLOUD", "connection.json"));

```

File: cloudnet-wrapper/src/main/java/de/dytanic/cloudnetwrapper/server/GameServer.java [hunk 19]
```diff
@@ -387,17 +317,131 @@ public boolean bootstrap() throws Exception {
...
+            Configuration section = configuration.getSection("server");
+            section.set("ip", CloudNetWrapper.getInstance().getWrapperConfig().getInternalIP());
+            section.set("port", serverProcess.getMeta().getPort());
+
+            maxPlayers = section.getInt("max-players");
+            motd = section.getString("motd");
+
+            configuration.set("server", section);
+            configuration.set("console.use-jline", false);
+            try (Writer outputStreamWriter = new OutputStreamWriter(Files.newOutputStream(Paths.get(path, "/config/glowstone.yml")),
```
Comment (GiantTreeLP) [inline cloudnet-wrapper/src/main/java/de/dytanic/cloudnetwrapper/server/GameServer.java]:
```suggestion
            try (Writer writer = new OutputStreamWriter(Files.newOutputStream(Paths.get(path, "config", "glowstone.yml")),
```

File: cloudnet-wrapper/src/main/java/de/dytanic/cloudnetwrapper/server/GameServer.java [hunk 20]
```diff
@@ -387,17 +317,131 @@ public boolean bootstrap() throws Exception {
...
+                                                                    StandardCharsets.UTF_8)) {
+                ConfigurationProvider.getProvider(YamlConfiguration.class).save(configuration, outputStreamWriter);
```
Comment (GiantTreeLP) [inline cloudnet-wrapper/src/main/java/de/dytanic/cloudnetwrapper/server/GameServer.java]:
```suggestion
                ConfigurationProvider.getProvider(YamlConfiguration.class).save(configuration, writer);
```

File: cloudnet-wrapper/src/main/java/de/dytanic/cloudnetwrapper/server/GameServer.java [hunk 21]
```diff
@@ -387,17 +317,131 @@ public boolean bootstrap() throws Exception {
...
+     *
+     * @return the finished server info of the configured server.
+     */
+    private ServerInfo configureNormalServer() {
+        String motd;
+        int maxPlayers;
+        Properties properties = new Properties();
+        try (InputStreamReader inputStreamReader = new InputStreamReader(Files.newInputStream(Paths.get(path,  "server.properties")))) {
+            try {
+                properties.load(inputStreamReader);
```
Comment (GiantTreeLP) [inline cloudnet-wrapper/src/main/java/de/dytanic/cloudnetwrapper/server/GameServer.java]:
```suggestion
        try (Reader reader = new InputStreamReader(Files.newInputStream(Paths.get(path,  "server.properties")))) {
            try {
                properties.load(reader);
```

File: cloudnet-wrapper/src/main/java/de/dytanic/cloudnetwrapper/server/GameServer.java [hunk 22]
```diff
@@ -550,7 +594,7 @@ private boolean templateDownloader() throws Exception {
     /**
      * Start the java process
      *
-     * @throws Exception
+     * @throws Exception Throws some exception, if something wrong on the start of a game server
```
Comment (GiantTreeLP) [inline cloudnet-wrapper/src/main/java/de/dytanic/cloudnetwrapper/server/GameServer.java:587]: Please check this definition again and be more specific

File: cloudnet-wrapper/src/main/java/de/dytanic/cloudnetwrapper/server/GameServer.java [hunk 23]
```diff
@@ -550,7 +594,7 @@ private boolean templateDownloader() throws Exception {
...
      */
     private void startProcess() throws Exception {
```
Comment (GiantTreeLP) [inline cloudnet-wrapper/src/main/java/de/dytanic/cloudnetwrapper/server/GameServer.java:589]: Is there no more specific exception being thrown?

**Review by GiantTreeLP — CHANGES_REQUESTED** [review]
In addition to the comments, please reformat the file properly.

File: cloudnet-wrapper/src/main/java/de/dytanic/cloudnetwrapper/server/GameServer.java (hunk 23 above)
Comment (TheMeinerLP) [reply cloudnet-wrapper/src/main/java/de/dytanic/cloudnetwrapper/server/GameServer.java:589]: Nope, alltime the same no specific.

File: cloudnet-wrapper/src/main/java/de/dytanic/cloudnetwrapper/server/GameServer.java (hunk 15 above)
Comment (TheMeinerLP) [reply cloudnet-wrapper/src/main/java/de/dytanic/cloudnetwrapper/server/GameServer.java]: But we have only four, but do all times the same thing

Comment (GiantTreeLP) [reply cloudnet-wrapper/src/main/java/de/dytanic/cloudnetwrapper/server/GameServer.java]: Then reverse the condition and only treat Glowstone specially.

**Review by GiantTreeLP — CHANGES_REQUESTED** [review]
Just fix the format, use the logger everywhere instead of System.out and we should be done.

File: cloudnet-wrapper/src/main/java/de/dytanic/cloudnetwrapper/server/GameServer.java (hunk 22 above)
Comment (TheMeinerLP) [reply cloudnet-wrapper/src/main/java/de/dytanic/cloudnetwrapper/server/GameServer.java:587]: Cannot make specific

**Comment by GiantTreeLP** [comment]
Please test these changes (including downloading plugins from the master and a url) and add documentation.

After this I am sure we can merge this

**Comment by TheMeinerLP** [comment]
Ok, i can do this

**Comment by TheMeinerLP** [comment]
Test Passed:
- [X] Download Plugin from URL(ProtocolSupport from CI)
- [X] Download Master Template(master-test template with protocolsupport plugin)
- [X] Download URL Template(Download from localhost nginx with h5ai)

Ready to merge @GiantTreeLP!

File: cloudnet-wrapper/src/main/java/de/dytanic/cloudnetwrapper/server/GameServer.java [hunk 24]
```diff
@@ -130,6 +132,93 @@ public Path getDir() {
         return dir;
     }
 
+    /**
+     * Download the {@link ServerInstallablePlugin] an copy to a web plugin cache.
```
Comment (derklaro) [inline cloudnet-wrapper/src/main/java/de/dytanic/cloudnetwrapper/server/GameServer.java]: I think here is a typo (an copy to a web...)

File: cloudnet-wrapper/src/main/java/de/dytanic/cloudnetwrapper/server/GameServer.java [hunk 25]
```diff
@@ -385,19 +312,135 @@ public boolean bootstrap() throws Exception {
...
-                                                                                               this.serverProcess.getMeta()));
-        System.out.println("Server " + this + " started in [" + (System.currentTimeMillis() - startupTime) + " milliseconds]");
-        this.startupTimeStamp = System.currentTimeMillis();
+    }
 
-        startProcess();
+    /**
+     * Configure a game server that's compatible with glowstone and set some properties.
+     *
+     * @return Given back a complete server info
```
Comment (derklaro) [inline cloudnet-wrapper/src/main/java/de/dytanic/cloudnetwrapper/server/GameServer.java:312]: Given back is redundant here

File: cloudnet-wrapper/src/main/java/de/dytanic/cloudnetwrapper/server/GameServer.java [hunk 26]
```diff
@@ -259,121 +262,45 @@ public boolean bootstrap() throws Exception {
...
-            properties.setProperty("server-port", serverProcess.getMeta().getPort() + NetworkUtils.EMPTY_STRING);
+        startProcess();
 
-            motd = properties.getProperty("motd");
-            try {
-                maxPlayers = Integer.parseInt(properties.getProperty("max-players"));
-            } catch (NumberFormatException e) {
-                maxPlayers = 100;
-            }
+        serverProcess.setServerStage(ServerStage.PROCESS);
```
Comment (derklaro) [inline cloudnet-wrapper/src/main/java/de/dytanic/cloudnetwrapper/server/GameServer.java]: I think if this is removed not a bag of rice will fall over in China (This is just set and then immediately changed again...)

**Review by derklaro — CHANGES_REQUESTED** [review]

File: cloudnet-wrapper/src/main/java/de/dytanic/cloudnetwrapper/server/GameServer.java (hunk 26 above)
Comment (TheMeinerLP) [reply cloudnet-wrapper/src/main/java/de/dytanic/cloudnetwrapper/server/GameServer.java]: Added to my to-do!

File: cloudnet-wrapper/src/main/java/de/dytanic/cloudnetwrapper/server/GameServer.java (hunk 25 above)
Comment (TheMeinerLP) [reply cloudnet-wrapper/src/main/java/de/dytanic/cloudnetwrapper/server/GameServer.java:312]: No, is used from parent method.

**Pull request merged**